src/passes/OptimizeInstructions.cpp

```diff
@@ -961,17 +984,30 @@ struct OptimizeInstructions
         }
       }
     } else if (auto* binary = boolean->dynCast<Binary>()) {
+      if (binary->op == SubInt32) {
+        if (auto* num = binary->left->dynCast<Const>()) {
+          if (num->value.geti32() == 0) {
+            // 0 - x   ==>   x
+            return binary->right;
+          }
+        }
+      }
       if (binary->op == OrInt32) {
         // an or flowing into a boolean context can consider each input as
         // boolean
         binary->left = optimizeBoolean(binary->left);
         binary->right = optimizeBoolean(binary->right);
       } else if (binary->op == NeInt32) {
-        // x != 0 is just x if it's used as a bool
         if (auto* num = binary->right->dynCast<Const>()) {
+          // x != 0 is just x if it's used as a bool
           if (num->value.geti32() == 0) {
             return binary->left;
           }
+          // x != -1   ==>    x ^ -1
```

Comment:
this comment should explain why this is worth doing, as the binary size is the same.

Reply:
Usually it cheaper for interpreters and first-tier compilers.

Reply:
@sunfishcode what do think? Is it make sense? It transform only for boolean contexts

Reply:
Btw LLVM and GCC also always canonize `x ^ -1` to `x != -1`. I'm wondering why `x != -1`
```asm
cmp   edi, -1
setne al
movzx eax, al
```

better than `x ^ -1`?

```asm
mov eax, edi
xor eax, 0xffffffff
```

Reply:
@kripken btw what do you think split OptimizeIntructions to two stages? It could be done via some parameter like `finalize`. Main idea is apply some transforms only once in last `optimization-instuctions` pass such as `x < 0` -> `x >>> 31`, `bool(x != 0)` -> `bool(x ^ -1)` and etc. Something like:
```js
add('ssa');
...other passes....
add('optimization-instuctions');
...other passes....
add('optimization-instuctions');
...other passes....
add('finalized-optimization-instuctions'); // last one
```

Reply:
If there's a reason we can add a new variation of it, and run it near the end, yeah. not sure there's a clear need for one yet though?

maybe let's split out this one controversial case from this PR, and investigate it separately.

Reply:
Sure. I'll remove this rule and open issue for discussing how better handle such of cases.

Reply:
LLVM's instcombine is optimizing for making other optimizations easy, rather then purely optimizing for speed. `x != -1` returns a value which is always 0 or 1, making it easier for other passes to know the set of values it might return. The idea is that if `x ^ -1` is cheaper in some context, the backend can easily pattern-match `x != -1` to produce it.

Reply:
Thank you! This was one of my assumptions.